src/software/ai/hl/stp/play/offensive_play_fsm.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include <include/boost/sml.hpp>
+
+#include "shared/constants.h"
+#include "shared/parameter/cpp_dynamic_parameters.h"
+#include "software/ai/hl/stp/play/offensive_play_fsm.h"
+#include "software/ai/hl/stp/play/play.h"
+#include "software/ai/hl/stp/tactic/attacker/attacker_tactic.h"
+#include "software/ai/hl/stp/tactic/move/move_tactic.h"
+#include "software/ai/hl/stp/tactic/receiver_tactic.h"
+#include "software/ai/passing/eighteen_zone_pitch_division.h"
+#include "software/ai/passing/pass_generator.h"
+#include "software/logger/logger.h"
+
+using Zones = std::unordered_set<EighteenZoneId>;
+
+struct OffensivePlayFSM
+{
+    class AttemptToShootWhileLookingForAPassState;
+    class TakePassState;
+    class StartState;
+
+    struct ControlParams
+    {
+        // number of additional offensive attacks on top of the minimum of 2 (passer and
+        // receiver)
+        unsigned int num_additional_offensive_tactics;
```

Comment:
we should provide the number of tactics always, not just for additional robots

i.e if the OffensiveFSM gets 1 tactic to assign, it should probably just return an attacker and obv not try to pass

thoughts?

Reply:
Something worth looking into is defining minimum (and maximum) number of tactics for a given play. In this case, the attacker only case is like a separate case in the FSM that would have to be carefully handled, i.e. make sure that a sole attacker is yielded correctly in every state, which seems annoying. What's the value of this?